connectors/sources/salesforce.py

```diff
@@ -0,0 +1,1443 @@
+            ]
+        )
+        query_builder.with_fields(queryable_docs_fields)
+        query_builder.with_fields(queryable_version_fields)
+        query_builder.with_fields(
+            [
+                "ContentDocument.Owner.Id",
+                "ContentDocument.Owner.Name",
+                "ContentDocument.Owner.Email",
+            ]
+        )
+        query_builder.with_fields(
+            [
+                "ContentDocument.CreatedBy.Id",
+                "ContentDocument.CreatedBy.Name",
+                "ContentDocument.CreatedBy.Email",
+            ]
+        )
+        query_builder.with_where(
+            f"ContentDocument.FileExtension IN ({where_in_clause})"
+        )
+
+        self._content_document_links_join = query_builder.build()
+        return self._content_document_links_join
+
+
+class SalesforceAPIToken:
+    def __init__(self, session, base_url, client_id, client_secret):
+        self._token = None
+        self.session = session
+        self.url = f"{base_url}{TOKEN_ENDPOINT}"
+        self.token_payload = {
+            "grant_type": "client_credentials",
+            "client_id": client_id,
+            "client_secret": client_secret,
+        }
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+    )
+    async def token(self):
+        if self._token:
+            return self._token
+
+        response_body = {}
+        try:
+            response = await self.session.post(self.url, data=self.token_payload)
+            response_body = await response.json()
+            response.raise_for_status()
+            self._token = response_body["access_token"]
+            return self._token
+        except ClientResponseError as e:
+            if 400 <= e.status < 500:
+                # 400s have detailed error messages in body
+                error_message = response_body.get("error", "No error dscription found.")
+                if error_message == "invalid_client":
+                    raise InvalidCredentialsException(
+                        f"The `client_id` and `client_secret` provided could not be used to generate a token. Status: {e.status}, message: {e.message}, details: {error_message}"
+                    ) from e
+                else:
+                    raise TokenFetchException(
+                        f"Could not fetch token from Salesforce: Status: {e.status}, message: {e.message}, details: {error_message}"
+                    ) from e
+            else:
+                raise TokenFetchException(
+                    f"Unexpected error while fetching Salesforce token. Status: {e.status}, message: {e.message}"
+                ) from e
+
+    def clear(self):
+        self._token = None
+
+
+class SalesforceSoqlBuilder:
+    def __init__(self, table):
+        self.table_name = table
+        self.fields = []
+        self.where = ""
+        self.order_by = ""
+        self.limit = ""
+
+    def with_id(self):
+        self.fields.append("Id")
```

Comment:
```suggestion
    def with_id(self):
        self.fields.append("Id")
        return self
```
I was expecting the methods in `SalesforceSoqlBuilder` will return `self`, so that it can be chained:

```python
query_builder = SalesforceSoqlBuilder("CaseFeed")
return query_builder.with_id().with_default_metafields().build()
```

Reply:
I didn't even know this was possible, thanks!